hnn_core/extracellular.py

```diff
@@ -0,0 +1,551 @@
+            length (in ms) to the data before plotting. Default: None
+        tmin : float | None
+            Start time of plot in milliseconds. If None, plot entire
+            simulation.
+        tmax : float | None
+            End time of plot in milliseconds. If None, plot entire simulation.
+        ax : instance of matplotlib figure | None
+            The matplotlib axis
+        decim : int | list of int | None (default)
+            Optional (integer) factor by which to decimate the raw dipole
+            traces. The SciPy function :func:`~scipy.signal.decimate` is used,
+            which recommends values <13. To achieve higher decimation factors,
+            a list of ints can be provided. These are applied successively.
+        color : string | array of floats | ``matplotlib.colors.ListedColormap``
+            The color to use for plotting (optional). The usual Matplotlib
+            standard color strings may be used (e.g., 'b' for blue). A color
+            can also be defined as an RGBA-quadruplet, or an array of
+            RGBA-values (one for each electrode contact trace to plot). An
+            instance of :class:`~matplotlib.colors.ListedColormap` may also be
+            provided.
+        voltage_offset : float | None (optional)
+            Amount to offset traces by on the voltage-axis. Useful for plotting
+            laminar arrays.
+        voltage_scalebar : float | None (optional)
+            Height, in units of uV, of a scale bar to plot in the top-left
+            corner of the plot.
+        contact_labels : list (optional)
+            Labels associated with the contacts to plot. Passed as-is to
+            :func:`~matplotlib.axes.Axes.set_yticklabels`.
+        show : bool
+            If True, show the figure
+
+        Returns
+        -------
+        fig : instance of plt.fig
+            The matplotlib figure handle into which time series were plotted.
+        """
+        from .viz import plot_extracellular
+
+        if trial_no is None:
+            plot_data = self.get_data()
+        elif isinstance(trial_no, (list, tuple, int, slice)):
+            plot_data = self.get_data()[trial_no, ]
+        else:
+            raise ValueError(f'unkown trial number type, got {trial_no}')
+
+        if isinstance(contact_no, (list, tuple, int, slice)):
+            plot_data = plot_data[:, contact_no, ]
+        elif contact_no is not None:
+            raise ValueError(f'unkown contact number type, got {contact_no}')
+
+        for trial_data in plot_data:
+            fig = plot_extracellular(
+                self.times, trial_data, self.sfreq, window_len=window_len,
+                tmin=tmin, tmax=tmax, ax=ax, decim=decim, color=color,
+                voltage_offset=voltage_offset,
+                voltage_scalebar=voltage_scalebar,
+                contact_labels=contact_labels,
+                show=show)
+        return fig
+
+
+class _ExtracellularArray:
```

Comment:
@cjayb I think we need to work on the readability of the code a bit. I did suggest separating the neuron objects ... but I think putting it in a separate class is a bit of an antipattern. We have had this with `NetworkBuilder` and it gets very hairy to track things as you're always referring to objects in the pure python class. 

Is there a way to do this separation in a single class? One way would be to declare the attributes as `None` in the init and then assign them to the `Neuron` object in a `record` method. I started attempting this but soon realized it's non-trivial for me as I'm not too familiar with the object. Perhaps easier for you? 

Also, init should be relatively thin. It should only contain assignments, not computations. Currently the `ExtracellularArray` init computes the `_transfer_resistance` etc

Reply:
Wouldn't making this a single classbreak MPI? I may be misunderstanding, but it seems like you'd run into a pickling problem (unless that is avoided by moving all the `Neuron` object instantiations to a separate method)

Reply:
> unless that is avoided by moving all the Neuron object instantiations to a separate method

Yeah that's exactly what I'm suggesting.

Reply:
I tried to include the neuron objects into `ExtracellularArray`, but ran into considerable problems. I simply don't think it's worth it! I need to understand why you consider this an "antipattern"? Why is it any different from `NetworkBuilder`, which is also a separate class?

> Currently the `ExtracellularArray` init computes the `_transfer_resistance` etc

Nope, but the neuron-containing `_ExtracellularArray` class does.

To clarify the above, in 07dec7b I've created `ExtracellularArrayBuilder`, which works very much like `NetworkBuilder`. This works. The problems I bumped into when trying to cram everything into a single class were no doubt due to some lingering Neuron pointers on multi-trial runs (the tests caught the errors beautifully). With the clean separation of a `Builder`-class, python garbage collection frees memory between trial invocations, and we're golden.

Reply:
Humm ... to be honest I don't like `NetworkBuilder` class but maybe I would like it after some clean up :) One issue I had when editing the code was that I could not tell where it was the builder class vs the pure python class from the variable name and attributes it accesses. Let me spend some more time with the code to see if the opinion changes!

Btw, did you ever get around to modifying `_clear_neuron_objects` to delete the pointers? Anyhow, perhaps we can live with this solution after some clean up